### PR TITLE
OpenCL: add prim & rev support for block, col, row, head, tail, num_elements and dims

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Adds prim & rev support for block-based functions: block, col, row, head, tail, num_elements and dims
Also: 
- renames the zero-based block() we had for OpenCL now to block_zero_based
- removes some rows/cols/size related checks from /opencl where we can just use prim/err ones.

## Tests

New functions have tests.

## Side Effects

/

## Release notes

OpenCL: Added prim & rev support for block-based functions: block, col, row, head, tail, num_elements and dims.

## Checklist

- [x] Math issue #1854 

- [x] Copyright holder: Rok Češnovar, Uni. of Ljubljana

### CODE REVIEW

**Comment by SteveBronder** [comment]
Nice! Should these also be added to `matrix_cl<>` as methods? Because then once all of the ones are added for [`vari_view`](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/vari.hpp#L218) we can hookup `var_value<matrix_cl<>>` to that and a bunch of the `rvalue`/`assign()` should work for matrix_cl<> types.

**Comment by rok-cesnovar** [comment]
Oh, I think Tadej was talking about that when adding transpose yeah. I would do those separately, this PR already touches 40 files. But yeah, that would be nice to have.

**Comment by SteveBronder** [comment]
Yeah it's totally fine and good to do those in a separate PR

**Comment by rok-cesnovar** [comment]
Ready for review.

File: stan/math/opencl/kernel_generator/block_zero_based.hpp [hunk 1]
```diff
@@ -283,15 +283,18 @@ class block_
 };
 
 /**
- * Block of a kernel generator expression.
+ * Block of a kernel generator expression. This version of zero-based in
+ * contrast to the 1-based block(), which is exposed to the Stan language.
```
Comment (t4c1) [inline stan/math/opencl/kernel_generator/block_zero_based.hpp:287]:
```suggestion
 * Block of a kernel generator expression. This version uses zero-based  indices in
 * contrast to the 1-based block(), which is exposed to the Stan language.
```

File: stan/math/opencl/prim/block.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * @param j Starting column.
+ * @param nrows Number of rows in block.
+ * @param ncols Number of columns in block.
+ * @throw std::out_of_range if either index is out of range.
+ */
+template <typename T_x,
+          typename = require_all_kernel_expressions_and_none_scalar_t<T_x>>
+inline auto block(T_x&& x, size_t i, size_t j, size_t nrows,
+                  size_t ncols) {  // NOLINT
+  return block_zero_based(std::forward<T_x>(x), i - 1, j - 1, nrows, ncols);
```
Comment (t4c1) [inline stan/math/opencl/prim/block.hpp:26]: What about aliasing? I think exposed functions should not have such issues. Maybe we should eval this.

File: stan/math/opencl/prim/col.hpp [hunk 3]
```diff
@@ -19,9 +19,9 @@ namespace math {
  * @throw std::invalid_argument if j is out of range.
  */
 template <typename T_x,
-          typename = require_all_kernel_expressions_and_none_scalar_t<T_x>>
+          typename = require_nonscalar_prim_or_rev_kernel_expression_t<T_x>>
 inline auto col(T_x&& x, size_t j) {  // NOLINT
-  return block(std::forward<T_x>(x), 0, j - 1, x.rows(), 1);
+  return block(std::forward<T_x>(x), 1, j, x.rows(), 1);
```
Comment (t4c1) [inline stan/math/opencl/prim/col.hpp:24]: Also change the include file to the new block.

File: stan/math/opencl/prim/head.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#ifndef STAN_MATH_OPENCL_PRIM_HEAD_HPP
+#define STAN_MATH_OPENCL_PRIM_HEAD_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
```
Comment (t4c1) [inline stan/math/opencl/prim/head.hpp:6]: You should include block, not kernel generator.

File: stan/math/opencl/prim/tail.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
+#ifndef STAN_MATH_OPENCL_PRIM_TAIL_HPP
+#define STAN_MATH_OPENCL_PRIM_TAIL_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
```
Comment (t4c1) [inline stan/math/opencl/prim/tail.hpp:6]: Include block, not kg.

File: stan/math/opencl/prim/tail.hpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * input) from the back of the specified vector or row vector.
+ *
+ * @tparam T_x type of input kernel generator expression.
+ * @param x input kernel generator expression.
+ * @param n Size of return.
+ * @return The first n elements of v.
+ * @throw std::out_of_range if n is out of range.
+ */
+template <typename T_x,
+          typename = require_nonscalar_prim_or_rev_kernel_expression_t<T_x>>
```
Comment (t4c1) [inline stan/math/opencl/prim/tail.hpp:23]:
```suggestion
          require_nonscalar_prim_or_rev_kernel_expression_t<T_x>* = nullptr>
```
Trying to make require usage consistent across all stan math.

File: stan/math/opencl/rev/block.hpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * @param j Starting column.
+ * @param nrows Number of rows in block.
+ * @param ncols Number of columns in block.
+ * @throw std::out_of_range if either index is out of range.
+ */
+template <typename T,
+          require_all_nonscalar_prim_or_rev_kernel_expression_t<T>* = nullptr,
+          require_any_var_t<T>* = nullptr>
+inline auto block(const T& m, size_t i, size_t j, size_t nrows, size_t ncols) {
+  var_value<matrix_cl<double>> res = block(m.val(), i, j, nrows, ncols);
```
Comment (t4c1) [inline stan/math/opencl/rev/block.hpp:27]: [optional] We could also use views.

File: stan/math/prim/err/check_column_index.hpp [hunk 8]
```diff
@@ -24,7 +24,9 @@ namespace math {
  * @param i column index to check
  * @throw <code>std::out_of_range</code> if index is an invalid column
  */
-template <typename T_y, typename = require_eigen_t<T_y>>
+template <typename T_y,
+          typename
+          = require_any_t<is_eigen<T_y>, is_prim_or_rev_kernel_expression<T_y>>>
```
Comment (t4c1) [inline stan/math/prim/err/check_column_index.hpp:29]:
```suggestion
          require_any_t<is_eigen<T_y>, is_prim_or_rev_kernel_expression<T_y>>* = nullptr>
```

File: stan/math/prim/err/check_matching_dims.hpp [hunk 9]
```diff
@@ -73,7 +75,11 @@ inline void check_matching_dims(const char* function, const char* name1,
  * @throw <code>std::invalid_argument</code> if the dimensions of the
  *    containers do not match
  */
-template <typename T1, typename T2, require_all_matrix_t<T1, T2>* = nullptr>
+template <typename T1, typename T2,
+          typename
+          = require_any_t<conjunction<is_matrix<T1>, is_matrix<T2>>,
+                          conjunction<is_prim_or_rev_kernel_expression<T1>,
+                                      is_prim_or_rev_kernel_expression<T2>>>>
```
Comment (t4c1) [inline stan/math/prim/err/check_matching_dims.hpp:82]:
```suggestion
          require_any_t<conjunction<is_matrix<T1>, is_matrix<T2>>,
                          conjunction<is_prim_or_rev_kernel_expression<T1>,
                                      is_prim_or_rev_kernel_expression<T2>>>* = nullptr>
```

File: stan/math/prim/err/check_square.hpp [hunk 10]
```diff
@@ -17,7 +17,9 @@ namespace math {
  * @param y Matrix to test
  * @throw <code>std::invalid_argument</code> if the matrix is not square
  */
-template <typename T_y, require_matrix_t<T_y>* = nullptr>
+template <typename T_y,
+          typename = require_any_t<is_matrix<T_y>,
+                                   is_prim_or_rev_kernel_expression<T_y>>>
```
Comment (t4c1) [inline stan/math/prim/err/check_square.hpp:22]:
```suggestion
          require_any_t<is_matrix<T_y>,
                                   is_prim_or_rev_kernel_expression<T_y>>* = nullptr>
```

File: stan/math/prim/err/check_vector.hpp [hunk 11]
```diff
@@ -22,7 +22,9 @@ namespace math {
  * @throw <code>std::invalid_argument</code> if x is not a row or column
  *   vector.
  */
-template <typename Mat, require_matrix_t<Mat>* = nullptr>
+template <typename Mat,
+          typename = require_any_t<is_matrix<Mat>,
+                                   is_prim_or_rev_kernel_expression<Mat>>>
```
Comment (t4c1) [inline stan/math/prim/err/check_vector.hpp:27]:
```suggestion
          require_any_t<is_matrix<Mat>,
                                   is_prim_or_rev_kernel_expression<Mat>>* = nullptr>
```

File: stan/math/prim/err/check_vector_index.hpp [hunk 12]
```diff
@@ -21,7 +21,9 @@ namespace math {
  * @param i row index to check
  * @throw <code>std::out_of_range</code> if the index is out of range.
  */
-template <typename T, typename = require_eigen_vector_t<T>>
+template <typename T,
+          typename = require_any_t<is_eigen_vector<T>,
+                                   is_prim_or_rev_kernel_expression<T>>>
```
Comment (t4c1) [inline stan/math/prim/err/check_vector_index.hpp:26]:
```suggestion
          require_any_t<is_eigen_vector<T>,
                                   is_prim_or_rev_kernel_expression<T>>* = nullptr>
```

File: stan/math/prim/fun/tail.hpp [hunk 13]
```diff
@@ -19,7 +19,7 @@ namespace math {
  * @return The last n elements of v.
  * @throw std::out_of_range if n is out of range.
  */
-template <typename T>
+template <typename T, typename = require_eigen_vector_t<T>>
```
Comment (t4c1) [inline stan/math/prim/fun/tail.hpp:22]:
```suggestion
template <typename T, require_eigen_vector_t<T>* = nullptr>
```

File: test/unit/math/opencl/rev/block_test.cpp [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+      = stan::math::to_matrix_cl(m3);
+  stan::math::matrix_v m2_res = stan::math::block(m2, 1, 2, 2, 1).eval();
+  stan::math::var_value<stan::math::matrix_cl<double>> m3_res_cl
+      = stan::math::block(m3_cl, 1, 2, 2, 1);
+  stan::math::var s = sum(m2_res) + sum(m3_res_cl);
+  s.grad();
+
+  EXPECT_MATRIX_NEAR(m2_res.val(), stan::math::from_matrix_cl(m3_res_cl.val()),
+                     1E-8);
+  EXPECT_MATRIX_NEAR(m2.adj(), m3.adj(), 1E-8);
```
Comment (t4c1) [inline test/unit/math/opencl/rev/block_test.cpp:40]: Why don't you use rev testing utility?

File: test/unit/math/opencl/rev/col_test.cpp [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+      = stan::math::to_matrix_cl(m3);
+  stan::math::matrix_v m2_res = stan::math::col(m2, 3);
+  stan::math::var_value<stan::math::matrix_cl<double>> m3_res_cl
+      = stan::math::col(m3_cl, 3);
+  stan::math::var s = sum(m2_res) + sum(m3_res_cl);
+  s.grad();
+
+  EXPECT_MATRIX_NEAR(m2_res.val(), stan::math::from_matrix_cl(m3_res_cl.val()),
+                     1E-8);
+  EXPECT_MATRIX_NEAR(m2.adj(), m3.adj(), 1E-8);
```
Comment (t4c1) [inline test/unit/math/opencl/rev/col_test.cpp:39]: Same question here.

File: test/unit/math/opencl/rev/head_test.cpp [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      = stan::math::to_matrix_cl(m3);
+  stan::math::matrix_v m2_res = stan::math::head(m2, 5);
+  stan::math::var_value<stan::math::matrix_cl<double>> m3_res_cl
+      = stan::math::head(m3_cl, 5);
+  stan::math::var s = sum(m2_res) + sum(m3_res_cl);
+  s.grad();
+
+  EXPECT_MATRIX_NEAR(m2_res.val(), stan::math::from_matrix_cl(m3_res_cl.val()),
+                     1E-8);
+  EXPECT_MATRIX_NEAR(m2.adj(), m3.adj(), 1E-8);
```
Comment (t4c1) [inline test/unit/math/opencl/rev/head_test.cpp:64]: and here

File: test/unit/math/opencl/rev/row_test.cpp [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      = stan::math::to_matrix_cl(m3);
+  stan::math::matrix_v m2_res = stan::math::row(m2, 3);
+  stan::math::var_value<stan::math::matrix_cl<double>> m3_res_cl
+      = stan::math::row(m3_cl, 3);
+  stan::math::var s = sum(m2_res) + sum(m3_res_cl);
+  s.grad();
+
+  EXPECT_MATRIX_NEAR(m2_res.val(), stan::math::from_matrix_cl(m3_res_cl.val()),
+                     1E-8);
+  EXPECT_MATRIX_NEAR(m2.adj(), m3.adj(), 1E-8);
```
Comment (t4c1) [inline test/unit/math/opencl/rev/row_test.cpp:39]: aaaand ... here.

File: test/unit/math/opencl/rev/tail_test.cpp [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+      = stan::math::to_matrix_cl(m3);
+  stan::math::matrix_v m2_res = stan::math::tail(m2, 5);
+  stan::math::var_value<stan::math::matrix_cl<double>> m3_res_cl
+      = stan::math::tail(m3_cl, 5);
+  stan::math::var s = sum(m2_res) + sum(m3_res_cl);
+  s.grad();
+
+  EXPECT_MATRIX_NEAR(m2_res.val(), stan::math::from_matrix_cl(m3_res_cl.val()),
+                     1E-8);
+  EXPECT_MATRIX_NEAR(m2.adj(), m3.adj(), 1E-8);
```
Comment (t4c1) [inline test/unit/math/opencl/rev/tail_test.cpp:74]: and here

**Review by t4c1 — CHANGES_REQUESTED** [review]

File: stan/math/opencl/prim/tail.hpp (hunk 6 above)
Comment (rok-cesnovar) [reply stan/math/opencl/prim/tail.hpp:23]: Agree.

File: stan/math/opencl/prim/block.hpp (hunk 2 above)
Comment (rok-cesnovar) [reply stan/math/opencl/prim/block.hpp:26]: Good call yeah.

Comment (t4c1) [reply stan/math/opencl/prim/block.hpp:26]: On the other hand exposed block() can not be assigned to. We only need to handle aliasing if/when we make opencl `assign` functions.

**Review by t4c1 — APPROVED** [review]

**Pull request merged**